_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,119 @@
+March 23, 2021. **The bug is present in all prior versions of Solidity.**
+
+We assigned the bug a severity level of "medium".
+
+## Technical Details
+
+**Summary**: The bytecode optimizer incorrectly re-used previously evaluated Keccak-256 hashes. You
+are unlikely to be affected if you do not compute Keccak-256 hashes in inline assembly.
+
+Solidity's bytecode optimizer has a step that can compute Keccak-256 hashes, if the contents of the
+memory, over which the Keccak-256 built-in function is invoked, are known during compilation time.
+This step also has a mechanism to determine that two Keccak-256 hashes are equal even if the values
+in memory are not known during compile time. This step had a bug where, the Keccak-256 hash of the
+same memory content, but of different sizes were considered equal. The following is a simple example
+that illustrates the bug:
+
+```solidity
+contract C {
+  function bug() public returns (uint a, uint b) {
+    assembly {
+      mstore(0, 0)
+      // The optimizer computes the value at compile time:
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      a := keccak256(0, 32)
+      // The optimizer incorrectly uses the cached value
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      // instead of 
+      // 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
+      b := keccak256(0, 23)
+    }
+  }
+}
+```
+
+In the above example, both the Keccak-256 hashes can be computed at compile time. Here are two other
+cases where the hashes cannot be computed at compile time, yet the optimizer considered them equal
+and thus modified the semantics:
+
+```solidity
+contract C {
+  function bug(string memory s) public returns (bool ret) {
+    assembly {
+      let a := keccak256(s, 32)
+      let b := keccak256(s, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+```solidity
+contract C {
+  function bug() public view returns (bool ret) {
+    assembly {
+      let x := calldataload(0)
+      mstore(0, x)
+      mstore(0x20, x)
+      let a := keccak256(0, 4)
+      // even though the memory location is different,
+      // the 32-byte content is the same.
+      let b := keccak256(0x20, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+Specifically, `keccak256(mpos1, length1)` and `keccak256(mpos2, length2)` in some cases were
+considered equal if `length1` and `length2`, when rounded up to nearest multiple of 32 were the
+same, and when the memory contents at `mpos1` and `mpos2` can be deduced to be equal.
+
+You may be **affected** if you compute multiple Keccak-256 hashes of the same content, but with
+different lengths inside inline assembly, and enabled the optimizer.
+
+### Unaffected Cases
+
```

Comment:
Maybe also add (could also put it further down or in a different section):

Apart from the keccak256 function, the compiler internally generates code that makes use of the keccak256 instruction in the following cases:
 - for indexed event arguments
 - for the function `abi.encodeWithSignature`
 - when using string or bytes keys in mappings
 - for accessing the data location of mappings and arrays

Reply:
Added